Return correct Retry After Header value

sven-rosenzweig opened the pull request:
If we do not want to suspend requests (max_sleep_time = 0), the lua scripts, calculating the rate limit, always returns 0 for the Retry-After Header value. Tools like terroform/opentofu will not slow down their deployment process, but instead simply abort

With this commit the lua scripts returns the correct value rather than 2 * max_sleep_time. Introducing a boolean indicating whether a request is suspended or not.

This commit adds three test cases for the rate limit calculus:
1. No limit reached,
2. Limit reached and suspension
3. Limit reached and no supspension

sven-rosenzweig committed:
Return correct Retry_After_Header value

If we do not want to suspend requests (max_sleep_time = 0), the lua
scripts, calculating the rate limit, always returns 0 for the
Retry-After Header value. Tools like terraform/OpenTofu will not slow
down their deployment process, but instead simply abort.

With this commit the lua scripts returns the correct value rather than
2 * max_sleep_time. Additionally we ensure never to return zero to the
client and log a warning if the lua script calculates zero as value.

This commit adds three test cases for the rate limit calculus:
1. No limit reached
2. Limit reached and suspension
3. Limit reached and no suspension

sven-rosenzweig committed:
Add Github workflow

sven-rosenzweig committed:
Remove flake8 errors

mutax commented on rate_limit/lua/redis_sliding_window.lua:
```diff
@@ -25,7 +25,7 @@ if remaining > 0 then
     redis.call('expire', key, window_seconds_int)
     -- Return the number of remaining requests. Retry after not relevant.
-    return {remaining, -1}
+    return {remaining, -1, false}
```
I _think_ the new value returned here is redundant.
The new value 'retry' will be false except in the case where remaining, the first value returned, is negative.
So it could be replaced in the calling python code by 
```python
  retry = remaining < 0
```
I see three places with a return. 
line 28 returns false, after we checked in line 22 that remaining is > 0
line 45 will return true. Here remaining  has to be <=0 and we substract 1, so the returned value is < 0
line 51 returns false, here remaining is set to 0.

sven-rosenzweig replied:
Nice review!

mutax reviewed (commented):
I have some trouble understanding the commit message of the rate limit fix and the PR descripton.

What I am missing is code that explicitly ensures that Retry-After is never set to 0. Need to check the math again - will add another review for that

mutax commented on rate_limit/lua/redis_sliding_window.lua:
```diff
@@ -42,9 +42,10 @@ if (retry_after_seconds < max_sleep_time_seconds_int) and (remaining - 1 >= -max
 -- Return if no more remaining requests and suspending request not possible.
 -- Ensure the 2nd argument (retry_after is greater than max_sleep_time_seconds_int)
-return {0, 2 * max_sleep_time_seconds_int}
+ --return {0, 2 * max_sleep_time_seconds_int}
```
the old code can be removed and the return seems to have an indent

felix-kaestner commented on rate_limit/lua/redis_sliding_window.lua:
```diff
@@ -42,9 +42,10 @@ if (retry_after_seconds < max_sleep_time_seconds_int) and (remaining - 1 >= -max
     redis.call('expire', key, window_seconds_int)
     -- Return if the request can be suspended.
-    return {remaining - 1 , retry_after_seconds}
+    return {remaining - 1 , retry_after_seconds, true}
```
```suggestion
    return {remaining - 1, retry_after_seconds, true}
```

toanju commented on .github/workflows/run-tox.yml:
```diff
@@ -0,0 +1,47 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python: ['3.10']
```
I guess https://github.com/sapcc/openstack-rate-limit-middleware/blob/master/setup.cfg is wrong and could be updated. In addition, if we use this code already in other python versions we could add these versions here.

sven-rosenzweig committed:
Update supported python versions

We do not support python version 2, so we remove it from the setup.cfg.

Run tox with multiple python versions.

mutax reviewed (approved):
lgtm
